### PR TITLE
FM2-422: Add support for observations with a location value

### PR DESCRIPTION
**Pull request by jnsereko**
## Description of what I changed
Add support for observations with a location value by returning a reference to a Location using Observation's valueText.  
cc @mozzy11 @ibacher 

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-422

## Checklist: I completed these to help reviewers :)

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Review by ibacher — CHANGES_REQUESTED** [review]
@jnsereko Thanks so much for taking this on!

So I think we can't do this translation directly in the `ObservationValueTranslatorImpl` class. The reason is that the extension we add is an extension on the observation itself. I think the code to create this extension looks something like this:

```java
if (existingObs.getValueText() != null && StringUtils.equals(existingObs.getComment(), "org.openmrs.Location")) {
	observation.addExtension(FhirConstants.OPENMRS_FHIR_EXT_OBS_LOCATION_VALUE, createLocationReferenceByUuid(existingObs.getValueText()));
}
```

`OPENMRS_FHIR_EXT_OBS_LOCATION_VALUE` should be constructed as `OPENMRS_FHIR_EXT_PREFIX + "/obs-location-value"`.

Sorry, I realise now that these are details I should've communicated in the original ticket.

File: api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/ObservationValueTranslatorImplTest.java [hunk 1]
```diff
@@ -48,6 +50,8 @@
 	
 	private static final String CONCEPT_VALUE_UUID = "12345-abcde-54321";
 	
+	private static final String LOCATION_UUID = "2321gh23-kj34h45-34jk3-34k34k";
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/translators/impl/ObservationValueTranslatorImplTest.java:53]: We probably should have a test at the `ObservationTranslatorImplTest` level anyway to verify that the Observation gets created properly.

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java [hunk 2]
```diff
@@ -45,7 +47,7 @@
 
 @Component
 @Setter(AccessLevel.PACKAGE)
-public class ObservationTranslatorImpl implements ObservationTranslator {
+public class ObservationTranslatorImpl extends BaseReferenceHandlingTranslator implements ObservationTranslator {
```
Comment (jnsereko) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java:50]: Thank you for your time @ibacher to look at this.
I am extending this here so as to access createLocationReferenceByUuid(....) method. I hope I have violated no rule.

Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java:50]: No! That's great!

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java [hunk 3]
```diff
@@ -126,6 +128,9 @@ public Observation toFhirResource(@Nonnull Obs observation) {
 			}
 		}
 		
+		if (observation.getValueText() != null && StringUtils.equals(observation.getComment(), "org.openmrs.Location")) {
+			obs.addExtension(FhirConstants.OPENMRS_FHIR_EXT_OBS_LOCATION_VALUE, createLocationReferenceByUuid(observation.getValueText()));
+		}
 		obs.setIssued(observation.getDateCreated());
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java:134]:
```suggestion
		}
		obs.setIssued(observation.getDateCreated());

```

File: api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java [hunk 4]
```diff
@@ -126,7 +128,11 @@ public Observation toFhirResource(@Nonnull Obs observation) {
...
+
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/translators/impl/ObservationTranslatorImpl.java:135]:
```suggestion
		}

		obs.setIssued(observation.getDateCreated());
```

**Review by ibacher — APPROVED** [review]

**Pull request merged**